### PR TITLE
0316 introduce client_attrs

### PR DESCRIPTION
**Pull request by zmstone**
Fixes https://emqx.atlassian.net/browse/EMQX-11882

Release version: v/e5.7.0

## Summary

Added `client_attrs` field to `clientinfo` object.
- `client_attrs` can be initialized with a regexp based extraction to set one `client_attrs.NAME` field.
- `client_attrs` can also be initialized from CONNECT packet's User-Property
- If there is a `client_attrs` field found in authentication results, it is merged to clientinfo `client_attrs` object
  - So far only HTTP auth and JWT auth supports extending `client_attrs`
- The placeholder `${client_attrs.NAME}` is allowed in authentication and authorization templates

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [x] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx/src/emqx_channel.erl [hunk 1]
```diff
@@ -1572,6 +1577,47 @@ enrich_client(ConnPkt, Channel = #channel{clientinfo = ClientInfo}) ->
...
+        extract_regexp := Regexp,
+        extract_as := AttrName
+    },
+    ClientInfo,
+    Peercert
+) when From =:= cn orelse From =:= dn ->
+    case extract_client_attr_from_cert(From, Regexp, Peercert) of
+        {ok, Value} ->
+            ?SLOG(
+                debug,
```
Comment (thalesmg) [inline apps/emqx/src/emqx_channel.erl:1592]: Could extracted values be sensitive?

File: apps/emqx/src/emqx_channel.erl [hunk 2]
```diff
@@ -1725,9 +1841,22 @@ do_authenticate(Credential, #channel{clientinfo = ClientInfo} = Channel) ->
...
 
-merge_auth_result(ClientInfo, AuthResult) when is_map(ClientInfo) andalso is_map(AuthResult) ->
-    IsSuperuser = maps:get(is_superuser, AuthResult, false),
-    maps:merge(ClientInfo, AuthResult#{is_superuser => IsSuperuser}).
+%% Merge authentication result into ClientInfo
+%% Authentication result may include:
+%% 1. `is_superuser': The superuser flag from various backends
+%% 2. `acl': ACL rules from JWT, HTTP auth backend
+%% 3. `client_attrs': Extra client attributes from JWT, HTTP auth backend
+%% 4. Maybe more non-standard fileds used by hook callbacks
```
Comment (thalesmg) [inline apps/emqx/src/emqx_channel.erl:1849]:
```suggestion
%% 4. Maybe more non-standard fields used by hook callbacks
```

File: changes/ce/feat-12750.en.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+attributes derived from the `clientinfo` fields upon accepting an MQTT connection.
+
+### Initialization of `client_attrs`
+
+- The `client_attrs` field can be initially populated based on the configuration from one of the
+  following sources:
+  - `clientid`: The MQTT client ID provided by the client.
+  - `username`: The username provided by the client.
+  - `cn`: The common name from the TLS client's certificate.
+  - `dn`: The distinguish name from the TLS client's certificate, i.e., the certificate "Subject".
```
Comment (thalesmg) [inline changes/ce/feat-12750.en.md:14]:
```suggestion
  - `dn`: The distinguished name from the TLS client's certificate, i.e., the certificate "Subject".
```

File: rel/i18n/emqx_schema.hocon [hunk 4]
```diff
@@ -1565,12 +1565,55 @@ This value specifies size of the batch.
...
+
+    EMQX uses durable storages to offload various data, such as MQTT messages, to disc."""
+
+client_attrs_init {
+  label: "Client attributes init"
+  desc: """~
+    Specify how to initialize client attributes.
+    One initial client attribute can be initialized as `client_attrs.NAME`,
+    where `NAME` is the name of the attribute specified in the config `extract_as`.
+    The initialized client attribute will be stored in the `client_attr` property with the specified name,
```
Comment (thalesmg) [inline rel/i18n/emqx_schema.hocon:1580]:
```suggestion
    The initialized client attribute will be stored in the `client_attrs` property with the specified name,
```

File: rel/i18n/emqx_schema.hocon [hunk 5]
```diff
@@ -1565,12 +1565,55 @@ This value specifies size of the batch.
...
+}
 
-EMQX uses durable storages to offload various data, such as MQTT messages, to disc."""
+client_attrs_init_extract_regex {
+  label: "Client attribute extract regex"
+  desc: """~
+    The regular expression to extract a client attribute from the client property specified by `client_attrs_init.extract_from` config.
+    The expression should match the entire client property value, and capturing groups are concatenated to make the client attribute.
+    For example if the client attribute is the first part of the client ID delemited by a dash, the regular expression would be `^(.+?)-.*$`.
+    Note that failure to match the regular expression will result in the client attribute being absence but not an empty string."""
```
Comment (thalesmg) [inline rel/i18n/emqx_schema.hocon:1608]:
```suggestion
    Note that failure to match the regular expression will result in the client attribute being absent but not an empty string."""
```

File: rel/i18n/emqx_schema.hocon [hunk 6]
```diff
@@ -1565,12 +1565,55 @@ This value specifies size of the batch.
...
+}
+
+client_attrs_init_extract_as {
+  label: "Name the extracted attribute"
+  desc: """~
+    The name of the client attribute extracted from the client property specified by `client_attrs_init.extract_from` config.
+    The extracted attribute will be stored in the `client_attr` property with this name.
```
Comment (thalesmg) [inline rel/i18n/emqx_schema.hocon:1615]:
```suggestion
    The extracted attribute will be stored in the `client_attrs` property with this name.
```

File: apps/emqx/src/emqx_channel.erl (hunk 1 above)
Comment (zmstone) [reply apps/emqx/src/emqx_channel.erl:1592]: not intended for secrets.

File: apps/emqx/src/emqx_channel.erl [hunk 7]
```diff
@@ -1572,6 +1575,47 @@ enrich_client(ConnPkt, Channel = #channel{clientinfo = ClientInfo}) ->
             {error, ReasonCode, Channel#channel{clientinfo = NClientInfo}}
     end.
 
+initialize_client_attrs_from_cert(
+    #{
```
Comment (savonarola) [inline apps/emqx/src/emqx_channel.erl:1579]: Do we purposely limit number of cert attrs to one?

File: apps/emqx/src/emqx_channel.erl [hunk 8]
```diff
@@ -1572,6 +1575,47 @@ enrich_client(ConnPkt, Channel = #channel{clientinfo = ClientInfo}) ->
...
+
+extract_client_attr_from_cert(cn, Regexp, Peercert) ->
+    CN = esockd_peercert:common_name(Peercert),
+    re_extract(CN, Regexp);
+extract_client_attr_from_cert(dn, Regexp, Peercert) ->
+    DN = esockd_peercert:subject(Peercert),
+    re_extract(DN, Regexp).
+
+re_extract(Str, Regexp) when is_binary(Str) ->
+    case re:run(Str, Regexp, [{capture, all_but_first, list}]) of
```
Comment (savonarola) [inline apps/emqx/src/emqx_channel.erl:1612]: According to the MQTT spec, most of the fields, like clientid, password, and user property, are UTF8 strings, so a user may expect REs to be UTF8-aware.
Currently, they are not; maybe we should at least add a warning?

File: apps/emqx/src/emqx_channel.erl (hunk 7 above)
Comment (zmstone) [reply apps/emqx/src/emqx_channel.erl:1579]: we can change the schema to allow many in the future if needed.

File: apps/emqx/src/emqx_channel.erl (hunk 8 above)
Comment (zmstone) [reply apps/emqx/src/emqx_channel.erl:1612]: you mean in the doc ?

Comment (savonarola) [reply apps/emqx/src/emqx_channel.erl:1612]: Yes, that in regexps we treat subjects as 1-byte strings.

Comment (zmstone) [reply apps/emqx/src/emqx_channel.erl:1612]: done adding doc in the last commit.

**Review by savonarola — APPROVED** [review]

**Comment by zmstone** [comment]
Follow up enhancements:
1. Introducing `variform`: #12851

**Comment by zmstone** [comment]
refactored in this PR: https://github.com/emqx/emqx/pull/12872

**Comment by emqxqa** [comment]
[Test Execution]()

**Pull request merged**